### PR TITLE
Support multiple channels for worker message passing at same function call depth

### LINKED ISSUE #35857: Runtime hangs when a function with sync or async send is called more than once
**Description:**

$title
Note this problem is not there if both strands are on the same thread.

**Steps to reproduce:**
```bal
public function main() {
    func();
    func();
}

function func() {
    @strand {thread: "any"}
    worker A {
        1 ->> B;
    }
    @strand {thread: "any"}
    worker B {
        int x1 = <- A;
    }
}
```

**Affected Versions:**
Ballerina 2201.0.2 (Swan Lake)

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

May be related to https://github.com/ballerina-platform/ballerina-lang/issues/35670, but here no NPE get thrown and this issue only seems to happen when strands are on different threads.

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
Fixes #35857

With this fix the below code will be supported.
```bal
import ballerina/test;

public function main() {
    foreach int i in 1...100 {
        bar(i, 2*i);
    }
}

function bar(int p, int q) {
    @strand {thread: "any"}
    worker A {
        p ->> B;
    }
    @strand {thread: "any"}
    worker B {
        int x1 = <- A;
        test:assertEquals(x1, p);
        q ->> function;
    }

    int x2 = <- B;
    test:assertEquals(x2, q);
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 1]
```diff
@@ -514,7 +522,12 @@ public static void loadChannelDetails(MethodVisitor mv, List<BIRNode.ChannelDeta
 
             mv.visitTypeInsn(NEW, CHANNEL_DETAILS);
             mv.visitInsn(DUP);
-            mv.visitLdcInsn(ch.name);
+            mv.visitVarInsn(ILOAD, invocationVarIndex);
+            mv.visitInvokeDynamicInsn(MAKE_CONCAT_WITH_CONSTANTS, "(I)Ljava/lang/String;",
+                    new Handle(H_INVOKESTATIC, "java/lang/invoke/StringConcatFactory", MAKE_CONCAT_WITH_CONSTANTS,
+                            "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;" +
+                                    "Ljava/lang/invoke/MethodType;Ljava/lang/String;[Ljava/lang/Object;)" +
+                                    "Ljava/lang/invoke/CallSite;", false), new Object[]{ch.name + ":\u0001"});
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:530]: Shall we move method signatures to the `JvmSignatures.java` class?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 2]
```diff
@@ -1082,21 +1089,33 @@ private void genFPCallIns(BIRTerminator.FPCall fpCall, String moduleClassName, B
             this.mv.visitMethodInsn(INVOKEVIRTUAL, FUNCTION_POINTER, "getFunction", GET_FUNCTION, false);
         }
 
+        boolean workerDerivative = fpCall.workerDerivative;
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:1092]: Do we need this `boolean` variable creation? We can directly pass the boolean argument.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 3]
```diff
@@ -1177,14 +1196,19 @@ private void loadTrueValueAsArg(int paramIndex) {
...
             this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "parent", GET_STRAND);
         }
         this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "wdChannels", GET_WD_CHANNELS);
-        this.mv.visitLdcInsn(ins.channel.value);
+        this.mv.visitVarInsn(ILOAD, invocationVarIndex);
+        this.mv.visitInvokeDynamicInsn(MAKE_CONCAT_WITH_CONSTANTS, "(I)Ljava/lang/String;",
+                new Handle(H_INVOKESTATIC, "java/lang/invoke/StringConcatFactory", MAKE_CONCAT_WITH_CONSTANTS,
+                        "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;" +
+                                "Ljava/lang/invoke/MethodType;Ljava/lang/String;[Ljava/lang/Object;)" +
+                                "Ljava/lang/invoke/CallSite;", false), new Object[]{ins.channel.value + ":\u0001"});
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:1211]: We can move the method signature constants to the `JvmSignature.java`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java (hunk 2 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:1092]: It is used twice, hence used a variable.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java [hunk 4]
```diff
@@ -437,6 +438,9 @@ public class JvmSignatures {
     public static final String XML_GET_ATTRIBUTE = "(L" + B_XML_QNAME + ";)L" + STRING_VALUE + ";";
     public static final String XML_GET_ITEM = "(I)L" + XML_VALUE + ";";
     public static final String XML_SET_ATTRIBUTE = "(L" + B_XML_QNAME + ";L" + B_STRING_VALUE + ";)V";
+    public static final String HANDLE_DESCRIPTOR_FOR_STRING_CONCAT = "(Ljava/lang/invoke/MethodHandles$Lookup;" +
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java:441]: Shall we use JvmConstants `OBJECT`, and `STRING_VALUE` with the concatenation?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 5]
```diff
@@ -174,10 +177,11 @@ public void genJMethodForBFunc(BIRFunction func, ClassWriter cw, BIRPackage modu
                                    BType attachedType, AsyncDataCollector asyncDataCollector) {
 
         BIRVarToJVMIndexMap indexMap = new BIRVarToJVMIndexMap();
-        indexMap.addIfNotExists(STRAND, symbolTable.stringType);
+        boolean isWorker = (func.flags & Flags.WORKER) == Flags.WORKER;
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:180]: Shall we use `Symbols.isFlagOn(func.flags, Flags.WORKER)`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 6]
```diff
@@ -188,9 +192,18 @@ public void genJMethodForBFunc(BIRFunction func, ClassWriter cw, BIRPackage modu
...
+        indexMap.addIfNotExists(STRAND, symbolTable.stringType);
+
         String funcName = func.name.value;
         BType retType = getReturnType(func);
-        String desc = JvmCodeGenUtil.getMethodDesc(func.type.paramTypes, retType);
+        String desc;
+        int invocationCountArgVarIndex = -1;
+        if (isWorker) {
+            invocationCountArgVarIndex = indexMap.addIfNotExists(INVOCATION_COUNT, symbolTable.stringType);
+            desc = JvmCodeGenUtil.getMethodDescForWorker(func.type.paramTypes, retType);
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:203]: As the `getMethodDescForWorker` method is only used here, we can include it in the line like this.
```java 
desc =  INITIAL_METHOD_DESC + "I" + populateMethodDesc(func.type.paramTypes) + generateReturnType(retType);
```

**Review by HindujaB — DISMISSED** [review]
LGTM

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 7]
```diff
@@ -514,7 +521,10 @@ public static void loadChannelDetails(MethodVisitor mv, List<BIRNode.ChannelDeta
 
             mv.visitTypeInsn(NEW, CHANNEL_DETAILS);
             mv.visitInsn(DUP);
-            mv.visitLdcInsn(ch.name);
+            mv.visitVarInsn(ILOAD, invocationVarIndex);
+            mv.visitInvokeDynamicInsn(MAKE_CONCAT_WITH_CONSTANTS, INT_TO_STRING,
+                    new Handle(H_INVOKESTATIC, STRING_CONCAT_FACTORY, MAKE_CONCAT_WITH_CONSTANTS,
+                            HANDLE_DESCRIPTOR_FOR_STRING_CONCAT, false), new Object[]{ch.name + ":\u0001"});
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:527]: Shall we add  ":\u0001" to JVM constants

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 8]
```diff
@@ -1202,14 +1227,17 @@ private void genWorkerSendIns(BIRTerminator.WorkerSend ins, int localVarOffset)
...
         this.mv.visitVarInsn(ALOAD, localVarOffset);
         if (!ins.isSameStrand) {
             this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "parent", GET_STRAND);
         }
         this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "wdChannels", GET_WD_CHANNELS);
-        this.mv.visitLdcInsn(ins.workerName.value);
+        this.mv.visitVarInsn(ILOAD, invocationVarIndex);
+        this.mv.visitInvokeDynamicInsn(MAKE_CONCAT_WITH_CONSTANTS, INT_TO_STRING,
+                new Handle(H_INVOKESTATIC, STRING_CONCAT_FACTORY, MAKE_CONCAT_WITH_CONSTANTS,
+                        HANDLE_DESCRIPTOR_FOR_STRING_CONCAT, false), new Object[]{ins.workerName.value + ":\u0001"});
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:1240]: ":\u0001" lets to move to constants

**Review by warunalakshitha — COMMENTED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**